### PR TITLE
statistical functions

### PR DESCRIPTION
**Pull request by izulin**
### Context

34 new statistical functions:  F.DIST, F.DIST.RT, F.INV, F.INV.RT, FDIST, FINV, WEIBULL, WEIBULL.DIST, HYPGEOMDIST, HYPGEOM.DIST, T.DIST, T.DIST.2T, T.DIST.RT, T.INV, T.INV.2T, TDIST, TINV, LOGNORM.DIST, LOGNORMDIST, LOGNORM.INV, LOGINV, NORM.DIST, NORMDIST, NORM.S.DIST, NORMSDIST, NORM.INV, NORMINV, NORM.S.INV, NORMSINV, PHI, NEGBINOM.DIST, NEGBINOMDIST, POISSON, POISSON.DIST.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1.
2.
3.

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation,
- [x] I described the modification in the CHANGELOG.md file.

### CODE REVIEW

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Added 18 mathematical functions ROMAN, ARABIC, FACT, FACTDOUBLE, COMBIN, COMBINA, GCD, LCM, MROUND, MULTINOMIAL, QUOTIENT, RANDBETWEEN, SERIESSUM, SIGN, SQRTPI, SUMX2MY2, SUMX2PY2, SUMXMY2.
 - Added 29 statistical functions EXPON.DIST, EXPONDIST, FISHER, FISHERINV, GAMMA, GAMMA.DIST, GAMMADIST, GAMMALN, GAMMALN.PRECISE, GAMMA.INV, GAMMAINV, GAUSS, BETA.DIST, BETADIST, BETA.INV, BETAINV, BINOM.DIST, BINOMDIST, BINOM.INV, BESSELI, BESSELJ, BESSELK, BESSELY, CHISQ.DIST, CHISQ.DIST.RT, CHISQ.INV, CHISQ.INV.RT, CHIDIST, CHIINV.
+- Added 32 statistical functions F.DIST, F.DIST.RT, F.INV, F.INV.RT, FDIST, FINV, WEIBULL, WEIBULL.DIST, HYPGEOMDIST, HYPGEOM.DIST, T.DIST, T.DIST.2T, T.DIST.RT, T.INV, T.INV.2T, TDIST, TINV, LOGNORM.DIST, LOGNORMDIST, LOGNORM.INV, LOGINV, NORM.DIST, NORMDIST, NORM.S.DIST, NORMSDIST, NORM.INV, NORMINV, NORM.S.INV, NORMSINV, PHI, NEGBINOM.DIST, NEGBINOMDIST.
```
[the remark this answers is not in the document]
Comment (izulin) [reply CHANGELOG.md:11]: Hmmm, right

File: test/interpreter/function-hypgeom.dist.spec.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      ['=HYPGEOM.DIST(12, 20, 12, 40, TRUE())'],
+      ['=HYPGEOM.DIST(12.1, 20, 12, 40, TRUE())'],
+      ['=HYPGEOM.DIST(4, 20, 4, 20, TRUE())'],
+      ['=HYPGEOM.DIST(4, 20, 4, 19.9, TRUE())'],
+      ['=HYPGEOM.DIST(4, 4, 20, 20, TRUE())'],
+      ['=HYPGEOM.DIST(4, 4, 20, 19.9, TRUE())'],
+      ['=HYPGEOM.DIST(10, 20, 20, 30, TRUE())'],
+      ['=HYPGEOM.DIST(10, 20.1, 20, 30, TRUE())'],
+      ['=HYPGEOM.DIST(0, 0.1, 0.1, 0.9, TRUE())'],
+      ['=HYPGEOM.DIST(10.9, 21, 20, 30.9, TRUE())'],
```
[the remark this answers is not in the document]
Comment (izulin) [reply test/interpreter/function-hypgeom.dist.spec.ts:79]: we test first arg for <= 0. We test second/third arg for <= first arg. We test fourth arg for <= second/third. There is no way to write test where second/third/fourth arg is <= 0 without violating those conditions or making first arg <= 0.

[the remark this answers is not in the document]
Comment (wojciechczerniak) [reply test/interpreter/function-hypgeom.dist.spec.ts:79]: 👍

**Review by wojciechczerniak — APPROVED** [review]
Thx 👍

File: test/interpreter/function-hypgeom.dist.spec.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(0.0000225475753840604, 6)
+    expect(engine.getCellValue(adr('A2'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A3'))).toEqual(1)
+    expect(engine.getCellValue(adr('A4'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueLarge))
```
Comment (bardek8) [inline test/interpreter/function-hypgeom.dist.spec.ts:85]: XL returns 1 here

File: test/interpreter/function-hypgeom.dist.spec.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    expect(engine.getCellValue(adr('A5'))).toEqual(1)
+    expect(engine.getCellValue(adr('A6'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueLarge))
```
Comment (bardek8) [inline test/interpreter/function-hypgeom.dist.spec.ts:87]: and here

File: test/interpreter/function-hypgeom.dist.spec.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    expect(engine.getCellValue(adr('A7'))).toEqual(1)
+    expect(engine.getCellValue(adr('A8'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueLarge))
+    expect(engine.getCellValue(adr('A9'))).toEqual(1)
+    expect(engine.getCellValue(adr('A10'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueLarge))
+    expect(engine.getCellValue(adr('A11'))).toBeCloseTo( 0.00614930629923134, 6)
+    expect(engine.getCellValue(adr('A12'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueLarge))
```
Comment (bardek8) [inline test/interpreter/function-hypgeom.dist.spec.ts:93]: XL accepts it

File: test/interpreter/function-negbinom.dist.spec.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      ['=NEGBINOM.DIST(0, 1, 0, FALSE())'],
+      ['=NEGBINOM.DIST(0, 1, -0.01, FALSE())'],
+      ['=NEGBINOM.DIST(0, 1, 1, FALSE())'],
+      ['=NEGBINOM.DIST(0, 1, 1.01, FALSE())'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.5)
+    expect(engine.getCellValue(adr('A2'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A3'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A4'))).toEqual(0)
```
Comment (bardek8) [inline test/interpreter/function-negbinom.dist.spec.ts:75]: XL returns error here

File: test/interpreter/function-negbinom.dist.spec.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    expect(engine.getCellValue(adr('A5'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A6'))).toEqual(1)
```
Comment (bardek8) [inline test/interpreter/function-negbinom.dist.spec.ts:77]: and here

File: test/interpreter/function-poisson.dist.spec.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(0.999999989952234, 6)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(0.983436391519386, 6)
+  })
+
+  it('checks bounds', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=POISSON.DIST(0, 0, FALSE())'],
+      ['=POISSON.DIST(-0.01, 0, FALSE())'],
+      ['=POISSON.DIST(0, -0.01, FALSE())'],
```
Comment (bardek8) [inline test/interpreter/function-poisson.dist.spec.ts:63]: XL accepts here

File: test/interpreter/function-t.dist.spec.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  })
+
+  it('checks bounds', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=T.DIST(-1, 1, FALSE())'],
+      ['=T.DIST(1, 0.9, FALSE())'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.159154942198517)
+    expect(engine.getCellValue(adr('A2'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
```
Comment (bardek8) [inline test/interpreter/function-t.dist.spec.ts:66]: interesting, XL says DIV/0 here

File: test/interpreter/function-t.inv.2t.spec.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    const engine = HyperFormula.buildFromArray([
+      ['=T.INV.2T(0.01, 1)'],
+      ['=T.INV.2T(0, 1)'],
+      ['=T.INV.2T(1, 1)'],
+      ['=T.INV.2T(1.01, 1)'],
+      ['=T.INV.2T(0.5, 0.9)'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(63.656741162866, 6)
+    expect(engine.getCellValue(adr('A2'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
```
Comment (bardek8) [inline test/interpreter/function-t.inv.2t.spec.ts:57]: XL returns DIV/0

File: test/interpreter/function-t.inv.2t.spec.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    expect(engine.getCellValue(adr('A3'))).toEqual(0)
+    expect(engine.getCellValue(adr('A4'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueLarge))
```
Comment (bardek8) [inline test/interpreter/function-t.inv.2t.spec.ts:59]: XL accepts

File: test/interpreter/function-t.inv.spec.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    const engine = HyperFormula.buildFromArray([
+      ['=T.INV(0.01, 1)'],
+      ['=T.INV(0, 1)'],
+      ['=T.INV(0.99, 1)'],
+      ['=T.INV(1, 1)'],
+      ['=T.INV(0.5, 0.9)'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(-31.820515953771, 6)
+    expect(engine.getCellValue(adr('A2'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
```
Comment (bardek8) [inline test/interpreter/function-t.inv.spec.ts:57]: DIV/0 in XL

File: test/interpreter/function-tdist.spec.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+      ['=TDIST(0, 1, 1.5)'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(0.5, 6)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(1, 6)
+    expect(engine.getCellValue(adr('A3'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A4'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A5'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A6'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A7'))).toEqualError(detailedError(ErrorType.NUM, ErrorMessage.IntegerExpected))
```
Comment (bardek8) [inline test/interpreter/function-tdist.spec.ts:76]: XL accepts

File: test/interpreter/function-weibull.dist.spec.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+  it('checks bounds', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=WEIBULL.DIST(0, 1, 1, FALSE())'],
+      ['=WEIBULL.DIST(-0.01, 0.01, 0.01, FALSE())'],
+      ['=WEIBULL.DIST(0, 0, 0.01, FALSE())'],
+      ['=WEIBULL.DIST(0, 0.01, 0, FALSE())'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(1)
```
Comment (bardek8) [inline test/interpreter/function-weibull.dist.spec.ts:59]: hm, XL returns 0 here, GS 1

**Review by bardek8 — CHANGES_REQUESTED** [review]
Thanks a lot!
Actually the only issue is with (in)compatibility with XL/GS, as they behave differently in the corner (out of bound) cases. You'll know best what is the desired behavior then, but I marked in comments all places that are different.

File: test/interpreter/function-hypgeom.dist.spec.ts [hunk 15]
```diff
@@ -0,0 +1,98 @@
+import {HyperFormula} from '../../src'
+import {ErrorType} from '../../src/Cell'
+import {ErrorMessage} from '../../src/error-message'
+import {adr, detailedError} from '../testUtils'
+
+describe('Function HYPGEOM.DIST', () => {
+  //In product #1, function takes 4 arguments.
```
Comment (bardek8) [inline test/interpreter/function-hypgeom.dist.spec.ts:7]: WDYT - in such situations the note above each different test would be too much clutter? The comment at the top can be easily missed

**Review by bardek8 — APPROVED** [review]
Thanks for the comments!

**Pull request merged**